File: codes/python/ch18/eigen_drift_diagnostic.py
#!/usr/bin/env python3
"""
eigen_drift_diagnostic.py

Chapter 18: Linear Spectral Eigenproblems
Computational Etude 18.5: build a spectrum lie detector.

Applies the drift-with-N diagnostic (spectrum_verify utility) to two
problems studied earlier in the chapter:

  (A) Dirichlet Laplacian u_xx + lambda u = 0 on (-1, 1), at N1 = 32
      and N2 = 48. Exact spectrum, well-ordered modes, ordinal matching
      sufficient.

  (B) Harmonic oscillator u_xx + (lambda - x^2) u = 0 on the real line,
      via rational_chebyshev with L = 4, at N1 = 32 and N2 = 48. Exact
      spectrum (Hermite), decay at infinity introduces the "infinite-
      interval tax".

For each problem we plot the reciprocal scaled drift 1/delta vs. mode
number j on a semilog axis (following Boyd Fig 7.7): trusted modes
appear at the TOP, suspect modes at the bottom. The cliff between the
two is the boundary of the trustworthy spectrum.

Reproduces the pedagogical lesson of Boyd Chapter 7.5: both the
finite-interval and unbounded problems show a sharp drop in reciprocal
drift ratio near the N/2 threshold.

Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
Part of "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

SCRIPT_DIR = Path(__file__).resolve().parent
sys.path.insert(0, str(SCRIPT_DIR))
sys.path.insert(0, str(SCRIPT_DIR.parent / "ch07"))
from cheb_matrix import cheb_matrix  # noqa: E402
from rational_chebyshev import rational_chebyshev_derivative_matrices  # noqa: E402
from spectrum_verify import verify_spectrum  # noqa: E402

plt.rcParams.update({
    "font.family": "serif",
    "font.serif": ["CMU Serif", "DejaVu Serif"],
    "mathtext.fontset": "cm",
    "font.size": 11,
    "axes.linewidth": 0.8,
    "xtick.major.width": 0.8,
    "ytick.major.width": 0.8,
    "figure.dpi": 150,
    "savefig.dpi": 300,
})

NAVY, SKY, CORAL, TEAL = "#142D6E", "#7896D2", "#E74C3C", "#16A085"

OUTPUT_DIR = SCRIPT_DIR.parent.parent.parent / "textbook" / "figures" / "ch18" / "python"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


def solve_laplacian(N: int) -> np.ndarray:
    D, _ = cheb_matrix(N)
    A = -(D @ D)[1:N, 1:N]
    return np.sort(np.linalg.eigvals(A).real)


def solve_oscillator(N: int, ell: float = 4.0) -> np.ndarray:
    _, D2, x = rational_chebyshev_derivative_matrices(N, ell)
    H = -D2 + np.diag(x ** 2)
    lam = np.linalg.eigvals(H)
    lam = lam.real[np.abs(lam.imag) < 1e-6]
    lam = lam[lam > 0]
    return np.sort(lam)


def plot_reciprocal_drift(ax, report, title, tol):
    j = np.arange(1, report.lam1.size + 1)
    # reciprocal drift: trusted modes at top
    inv_ord = 1.0 / np.maximum(report.delta_ordinal, 1e-18)
    inv_nst = 1.0 / np.maximum(report.delta_nearest, 1e-18)
    ax.semilogy(j, inv_ord, "o", color=NAVY, markersize=4,
                markerfacecolor="white", markeredgewidth=1.0,
                label=r"$1/\delta_{j,\mathrm{ordinal}}$")
    ax.semilogy(j, inv_nst, "x", color=CORAL, markersize=5,
                markeredgewidth=1.0,
                label=r"$1/\delta_{j,\mathrm{nearest}}$")
    ax.axhline(1.0 / tol, color="k", linestyle="--", linewidth=0.6,
               alpha=0.5)
    ax.text(1, 2.0 / tol, fr"$1/$tol $= 1/{tol:.0e}$",
            fontsize=8, color="k", alpha=0.7)
    ax.set_title(title, fontsize=10)
    ax.set_xlabel("mode number $j$")
    ax.set_ylabel(r"$1 / \delta_j$  (trusted = high)")
    ax.grid(True, which="both", alpha=0.25, linewidth=0.4)
    ax.legend(loc="upper right", fontsize=9)


def make_figure(N1: int = 32, N2: int = 48, tol: float = 1e-3):
    # Problem A: Dirichlet Laplacian
    lam1_A = solve_laplacian(N1)
    lam2_A = solve_laplacian(N2)
    rep_A = verify_spectrum(lam1_A, lam2_A, tol=tol)

    # Problem B: Oscillator on the real line
    lam1_B = solve_oscillator(N1)
    lam2_B = solve_oscillator(N2)
    rep_B = verify_spectrum(lam1_B, lam2_B, tol=tol)

    fig, axes = plt.subplots(1, 2, figsize=(11.0, 4.5))
    plot_reciprocal_drift(axes[0], rep_A,
        rf"Dirichlet Laplacian,  $N_1 = {N1}$, $N_2 = {N2}$", tol)
    plot_reciprocal_drift(axes[1], rep_B,
        rf"Harmonic oscillator (real line),  $N_1 = {N1}$, $N_2 = {N2}$", tol)
    fig.tight_layout()
    fig.savefig(OUTPUT_DIR / "eigen_drift_diagnostic.pdf", bbox_inches="tight", pad_inches=0.05)
    fig.savefig(OUTPUT_DIR / "eigen_drift_diagnostic.png", bbox_inches="tight", pad_inches=0.05)
    plt.close(fig)

    return {
        "N1": N1, "N2": N2, "tol": tol,
        "laplacian": {
            "n_trusted":       int(rep_A.n_trusted),
            "lam1":            rep_A.lam1.tolist(),
            "lam2":            rep_A.lam2.tolist(),
            "delta_ordinal":   rep_A.delta_ordinal.tolist(),
            "delta_nearest":   rep_A.delta_nearest.tolist(),
            "trusted":         rep_A.trusted.astype(int).tolist(),
        },
        "oscillator": {
            "n_trusted":       int(rep_B.n_trusted),
            "lam1":            rep_B.lam1.tolist(),
            "lam2":            rep_B.lam2.tolist(),
            "delta_ordinal":   rep_B.delta_ordinal.tolist(),
            "delta_nearest":   rep_B.delta_nearest.tolist(),
            "trusted":         rep_B.trusted.astype(int).tolist(),
        },
    }


def main():
    p = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    p.add_argument("--N1",  type=int,   default=32)
    p.add_argument("--N2",  type=int,   default=48)
    p.add_argument("--tol", type=float, default=1e-3)
    p.add_argument("--dump", type=str,  default=None)
    args = p.parse_args()

    r = make_figure(N1=args.N1, N2=args.N2, tol=args.tol)
    print(f"[Etude 18.5]  drift diagnostic at N1 = {args.N1}, N2 = {args.N2}, tol = {args.tol}")
    print(f"  Dirichlet Laplacian : trusted = {r['laplacian']['n_trusted']} of {len(r['laplacian']['lam1'])}")
    print(f"  Harmonic oscillator  : trusted = {r['oscillator']['n_trusted']} of {len(r['oscillator']['lam1'])}")
    print(f"  figure: {OUTPUT_DIR / 'eigen_drift_diagnostic.pdf'}")
    if args.dump:
        Path(args.dump).write_text(json.dumps(r, indent=2))
        print(f"  dumped: {args.dump}")


if __name__ == "__main__":
    main()

File: codes/julia/ch18/eigen_drift_diagnostic.jl
# eigen_drift_diagnostic.jl
#
# Chapter 18: Linear Spectral Eigenproblems
# Computational Etude 18.5: build a spectrum lie detector.
#
# Applies the drift-with-N diagnostic (spectrum_verify) to:
#   (A) Dirichlet Laplacian u_xx + lambda u = 0 on (-1, 1)
#   (B) Harmonic oscillator u_xx + (lambda - x^2) u = 0 on the real line.
#
# Reproduces Boyd (2000) Fig 7.7.
#
# Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
# Part of "Computational Etudes: A Spectral Approach"

using CairoMakie
using Colors
using LinearAlgebra
using Printf

include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))
include(joinpath(@__DIR__, "rational_chebyshev.jl"))
include(joinpath(@__DIR__, "spectrum_verify.jl"))
using .RationalChebyshev: rational_chebyshev_derivative_matrices
using .SpectrumVerify: verify_spectrum, SpectrumReport

_json_val(x::Bool)          = x ? "true" : "false"
_json_val(x::Integer)       = string(x)
_json_val(x::Real)          = isfinite(x) ? string(Float64(x)) : (isnan(x) ? "NaN" : (x > 0 ? "Infinity" : "-Infinity"))
_json_val(x::AbstractString)= "\"" * replace(String(x), "\\" => "\\\\", "\"" => "\\\"") * "\""
_json_val(v::AbstractVector)= "[" * join(_json_val.(v), ", ") * "]"
function _json_val(d::AbstractDict)
    pairs = String[]
    for (k, v) in d; push!(pairs, _json_val(string(k)) * ": " * _json_val(v)); end
    return "{" * join(pairs, ", ") * "}"
end
write_json(path, x) = open(io -> print(io, _json_val(x)), path, "w")

set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 11,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 9,),
))

const NAVY  = colorant"#142D6E"
const CORAL = colorant"#E74C3C"

const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch18", "julia")
mkpath(OUTPUT_DIR)

function solve_laplacian(N::Int)
    D, _ = cheb_matrix(N)
    A = -(D * D)[2:N, 2:N]
    return sort(real.(eigvals(A)))
end

function solve_oscillator(N::Int, ell::Real = 4.0)
    _, D2, x = rational_chebyshev_derivative_matrices(N, ell)
    H = -D2 + Diagonal(x .^ 2)
    lam = eigvals(H)
    keep = abs.(imag.(lam)) .< 1e-6
    lam = real.(lam[keep])
    lam = lam[lam .> 0]
    return sort(lam)
end

function plot_panel!(ax, rep::SpectrumReport, title_str::String, tol::Real)
    j = collect(1:length(rep.lam1))
    inv_ord = 1.0 ./ max.(rep.delta_ordinal, 1e-18)
    inv_nst = 1.0 ./ max.(rep.delta_nearest, 1e-18)
    scatter!(ax, j, inv_ord; color = :white, strokecolor = NAVY,
             strokewidth = 1.0, markersize = 7,
             label = "1/δ_ordinal")
    scatter!(ax, j, inv_nst; color = CORAL, marker = :xcross,
             markersize = 9,
             label = "1/δ_nearest")
    hlines!(ax, [1.0 / tol]; color = :black, linestyle = :dash,
            linewidth = 0.6, alpha = 0.5)
    ax.title[] = title_str
end

function make_figure(; N1::Int = 32, N2::Int = 48, tol::Real = 1e-3,
                     dump_path = nothing)
    lam1_A = solve_laplacian(N1);  lam2_A = solve_laplacian(N2)
    rep_A  = verify_spectrum(lam1_A, lam2_A; tol = tol)

    lam1_B = solve_oscillator(N1); lam2_B = solve_oscillator(N2)
    rep_B  = verify_spectrum(lam1_B, lam2_B; tol = tol)

    fig = Figure(size = (1100, 450))

    ax1 = Axis(fig[1, 1];
        xlabel = "mode number j", ylabel = "1 / δ_j  (trusted = high)",
        yscale = log10,
        title = "Dirichlet Laplacian,  N₁ = $N1, N₂ = $N2")
    plot_panel!(ax1, rep_A, "Dirichlet Laplacian,  N₁ = $N1, N₂ = $N2", tol)
    axislegend(ax1, position = :rt, labelsize = 9, patchsize = (12, 12))

    ax2 = Axis(fig[1, 2];
        xlabel = "mode number j", ylabel = "1 / δ_j  (trusted = high)",
        yscale = log10,
        title = "Harmonic oscillator (real line),  N₁ = $N1, N₂ = $N2")
    plot_panel!(ax2, rep_B, "Harmonic oscillator,  N₁ = $N1, N₂ = $N2", tol)
    axislegend(ax2, position = :rt, labelsize = 9, patchsize = (12, 12))

    save(joinpath(OUTPUT_DIR, "eigen_drift_diagnostic.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "eigen_drift_diagnostic.png"), fig, px_per_unit = 4)

    r = Dict{String,Any}(
        "N1" => N1, "N2" => N2, "tol" => tol,
        "laplacian"  => Dict(
            "n_trusted"     => rep_A.n_trusted,
            "lam1"          => rep_A.lam1,
            "lam2"          => rep_A.lam2,
            "delta_ordinal" => rep_A.delta_ordinal,
            "delta_nearest" => rep_A.delta_nearest,
            "trusted"       => Int.(rep_A.trusted),
        ),
        "oscillator" => Dict(
            "n_trusted"     => rep_B.n_trusted,
            "lam1"          => rep_B.lam1,
            "lam2"          => rep_B.lam2,
            "delta_ordinal" => rep_B.delta_ordinal,
            "delta_nearest" => rep_B.delta_nearest,
            "trusted"       => Int.(rep_B.trusted),
        ),
    )
    dump_path !== nothing && write_json(dump_path, r)
    return r, rep_A, rep_B
end

function parse_args()
    N1 = 32; N2 = 48; tol = 1e-3; dump_path = nothing; i = 1
    while i <= length(ARGS)
        if ARGS[i] == "--N1" && i < length(ARGS); N1 = parse(Int, ARGS[i+1]); i += 2
        elseif ARGS[i] == "--N2" && i < length(ARGS); N2 = parse(Int, ARGS[i+1]); i += 2
        elseif ARGS[i] == "--tol" && i < length(ARGS); tol = parse(Float64, ARGS[i+1]); i += 2
        elseif ARGS[i] == "--dump" && i < length(ARGS); dump_path = ARGS[i+1]; i += 2
        else i += 1 end
    end
    return N1, N2, tol, dump_path
end

if abspath(PROGRAM_FILE) == @__FILE__
    N1, N2, tol, dp = parse_args()
    _, rA, rB = make_figure(; N1 = N1, N2 = N2, tol = tol, dump_path = dp)
    @printf("[Etude 18.5]  drift diagnostic at N1 = %d, N2 = %d, tol = %g\n", N1, N2, tol)
    @printf("  Dirichlet Laplacian : trusted = %d of %d\n", rA.n_trusted, length(rA.lam1))
    @printf("  Harmonic oscillator  : trusted = %d of %d\n", rB.n_trusted, length(rB.lam1))
    println("  figure: ", joinpath(OUTPUT_DIR, "eigen_drift_diagnostic.pdf"))
    dp !== nothing && println("  dumped: ", dp)
end
